proof/refine/ARM/Tcb_R.thy

```diff
@@ -1509,741 +1783,963 @@ lemma tc_corres_caps:
+  apply (clarsimp simp: obj_at'_def projectKO_eq projectKO_sc)
+  done
+
+abbreviation all_invs'_but_valid_release_queue where
+  "all_invs'_but_valid_release_queue \<equiv>
+   \<lambda>s. valid_pspace' s \<and> sch_act_wf (ksSchedulerAction s) s
+                       \<and> valid_queues s
+                       \<and> sym_refs (list_refs_of_replies' s)
+                       \<and> if_live_then_nonz_cap' s \<and> if_unsafe_then_cap' s
+                       \<and> valid_idle' s
+                       \<and> valid_global_refs' s \<and> valid_arch_state' s
+                       \<and> valid_irq_node' (irq_node' s) s
+                       \<and> valid_irq_handlers' s
+                       \<and> valid_irq_states' s
+                       \<and> valid_machine_state' s
+                       \<and> irqs_masked' s
+                       \<and> valid_queues' s
+                       \<and> ct_not_inQ s
+                       \<and> ct_idle_or_in_cur_domain' s
+                       \<and> valid_pde_mappings' s
+                       \<and> pspace_domain_valid s
+                       \<and> ksCurDomain s \<le> maxDomain
+                       \<and> valid_dom_schedule' s
+                       \<and> untyped_ranges_zero' s
+                       \<and> cur_tcb' s"
+
+lemma threadSet_tcbInReleaseQueue_invs':
+  "\<lbrace>\<lambda>s. all_invs'_but_valid_release_queue s \<and>
+        (\<forall>tcb. ko_at' tcb tptr s \<longrightarrow>
+                valid_release_queue_iff
+                  (s\<lparr>ksPSpace := ksPSpace s(tptr \<mapsto> KOTCB (tcbInReleaseQueue_update (\<lambda>_. v) tcb))\<rparr>))\<rbrace>
+   threadSet (tcbInReleaseQueue_update (\<lambda>_. v)) tptr
+   \<lbrace>\<lambda>_. invs'\<rbrace>"
+  apply (clarsimp simp add: invs'_def valid_state'_def split del: if_split)
+  apply (wp threadSet_valid_pspace'T threadSet_sch_actT_P[where P=False, simplified]
+           threadSet_iflive'T threadSet_ifunsafe'T threadSet_idle'T threadSet_not_inQ
+           valid_irq_node_lift valid_irq_handlers_lift'' threadSet_ct_idle_or_in_cur_domain'
+           threadSet_cur untyped_ranges_zero_lift threadSet_valid_queues threadSet_valid_queues'
+         | rule refl threadSet_wp [THEN hoare_vcg_conj_lift]
+         | clarsimp simp: tcb_cte_cases_def cteCaps_of_def)+
+  apply (fastforce simp: obj_at'_def projectKO_eq projectKO_tcb objBitsKO_def inQ_def comp_def )
+  done
+
+lemma getScTime_wp:
+  "\<lbrace>\<lambda>s. \<forall>rt. (\<forall>tcb p sc. (ko_at' tcb t s \<and> tcbSchedContext tcb = Some p \<and> ko_at' sc p s
+                                                          \<longrightarrow> rTime (refillHd sc) = rt)) \<longrightarrow> Q rt s\<rbrace>
+   getScTime t \<lbrace>Q\<rbrace>"
+  by (wpsimp wp: threadGet_wp simp: getScTime_def getTCBSc_def obj_at'_def)
+
+schematic_goal mapM_getScTime_wp:
+  "\<lbrace>?P\<rbrace> mapM getScTime ts \<lbrace>Q\<rbrace>"
```

Comment:
It's a bit unusual to have a specific `mapM` lemma for a constant, but no problem if it makes other proofs easier.